# Credit Card Fraud Detection

***By Joe Corliss***

[Data source](https://www.kaggle.com/mlg-ulb/creditcardfraud)

I'm going to try to keep this kernel short and sweet, just to test out a few models.

This is my second machine learning project on Kaggle. I would really appreciate any feedback!

## Table of Contents

1. [Check the Data](#1)
2. [Pre-processing](#2)
3. [Predictive Modeling](#3)
    1. [Linear Classification with Stochastic Gradient Descent (SGD)](#3.1)
    2. [Random Forest Classifier](#3.2)

# Check the Data
<a id='1'></a>

In [1]:
import numpy as np
import pandas as pd

Read in the data to a pandas dataframe.

In [2]:
df = pd.read_csv('creditcard.csv', low_memory=False)

Check basic metadata.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
Time      284807 non-null float64
V1        284807 non-null float64
V2        284807 non-null float64
V3        284807 non-null float64
V4        284807 non-null float64
V5        284807 non-null float64
V6        284807 non-null float64
V7        284807 non-null float64
V8        284807 non-null float64
V9        284807 non-null float64
V10       284807 non-null float64
V11       284807 non-null float64
V12       284807 non-null float64
V13       284807 non-null float64
V14       284807 non-null float64
V15       284807 non-null float64
V16       284807 non-null float64
V17       284807 non-null float64
V18       284807 non-null float64
V19       284807 non-null float64
V20       284807 non-null float64
V21       284807 non-null float64
V22       284807 non-null float64
V23       284807 non-null float64
V24       284807 non-null float64
V25       284807 non-null float64
V26  

Check the first few rows of the data.

In [5]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...         V21       V22       V23       V24  \
0  0.098698  0.363787  ...   -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425  ...   -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  ...    0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024  ...   -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  ...   -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4 -0.206010  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

Get descriptive statistics.

In [6]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  3.919560e-15  5.688174e-16 -8.769071e-15  2.782312e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean  -1.552563e-15  2.010663e-15 -1.694249e-15 -1.927028e-16 -3.137024e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

           ...                 V21           V22           V23           V24  \
count      ...        2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean       ...        1.537294e-16  7.959909e-16  5.367590e-16  4.458112e-15   
std        ...        7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min        ...       -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%        ...       -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%        ...       -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%        ...        1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max        ...        2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   1.453003e-15  1.699104e-15 -3.660161e-16 -1.206049e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

Note that we will need to standardize the data (rescale each column to zero mean and unit variance) to aid learning algorithms.

How balanced are the classes, i.e. how common are fraudulent transactions?

In [7]:
df['Class'].value_counts() / df.shape[0]

0    0.998273
1    0.001727
Name: Class, dtype: float64

Only about 0.173% of transactions are fraudulent.

Is there any missing data?

In [8]:
df.isnull().sum().sum()

0

No. Great!

# Pre-processing
<a id='2'></a>

Split the data into features and target variable.

In [9]:
X = df.drop(columns=['Class']) # Features
y = df['Class'] # Target
df = None # Delete original dataframe to save memory

## Train/test split

In [10]:
from sklearn.model_selection import train_test_split

Do an 80/20 train/test split.

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

## Standardization

In [12]:
from sklearn.preprocessing import StandardScaler

Fit the scaler only to the training data to avoid test set leakage.

In [13]:
scaler = StandardScaler().fit(X_train)

Scale the data.

In [14]:
X_train = pd.DataFrame(scaler.transform(X_train), columns=X_train.columns)
X_test  = pd.DataFrame(scaler.transform(X_test),  columns=X_test.columns)

# Predictive Modeling
<a id='3'></a>

In [15]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import matthews_corrcoef, make_scorer, average_precision_score, confusion_matrix, classification_report

## Linear Classification with Stochastic Gradient Descent (SGD)
<a id='3.1'></a>

In [16]:
from sklearn.linear_model import SGDClassifier

The estimator `SGDClassifier` ([documentation](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html)) implements multiple linear classifiers with SGD training, which makes it very fast on large datasets at the cost of accuracy.

### Train with grid search

The linear classifier is chosen with the `loss` hyperparameter. We'll try a linear SVC (`loss = 'hinge'`) and logistic regression (`loss = 'log'`).

In [22]:
param_grid = [{'loss': ['hinge'],
               'alpha': [10.0**k for k in range(-6,4)],
               'max_iter': [1000],
               'tol': [1e-3],
               'class_weight': [None, 'balanced'],
               'warm_start': [True]},
              {'loss': ['log'],
               'penalty': ['l2', 'l1'],
               'alpha': [10.0**k for k in range(-6,4)],
               'max_iter': [1000],
               'tol': [1e-3],
               'warm_start': [True]}]

We'll use the Matthews correlation coefficient (MCC) ([documentation](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html)) as our scoring metric.

In [23]:
grid = GridSearchCV(estimator=SGDClassifier(), param_grid=param_grid, scoring=make_scorer(matthews_corrcoef), 
                    n_jobs=4, verbose=3, return_train_score=True)

Perform the grid search:

In [24]:
grid.fit(X_train, y_train)

Fitting 3 folds for each of 40 candidates, totalling 120 fits


[Parallel(n_jobs=4)]: Done  24 tasks      | elapsed:   10.8s
[Parallel(n_jobs=4)]: Done 120 out of 120 | elapsed:   46.9s finished


GridSearchCV(cv=None, error_score='raise',
       estimator=SGDClassifier(alpha=0.0001, average=False, class_weight=None, epsilon=0.1,
       eta0=0.0, fit_intercept=True, l1_ratio=0.15,
       learning_rate='optimal', loss='hinge', max_iter=None, n_iter=None,
       n_jobs=1, penalty='l2', power_t=0.5, random_state=None,
       shuffle=True, tol=None, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=4,
       param_grid=[{'loss': ['hinge'], 'alpha': [1e-06, 1e-05, 0.0001, 0.001, 0.01, 0.1, 1.0, 10.0, 100.0, 1000.0], 'max_iter': [1000], 'tol': [0.001], 'class_weight': [None, 'balanced'], 'warm_start': [True]}, {'loss': ['log'], 'penalty': ['l2', 'l1'], 'alpha': [1e-06, 1e-05, 0.0001, 0.001, 0.01, 0.1, 1.0, 10.0, 100.0, 1000.0], 'max_iter': [1000], 'tol': [0.001], 'warm_start': [True]}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=make_scorer(matthews_corrcoef), verbose=3)

What are the optimal parameters from our chosen grid?

In [26]:
grid.best_params_

{'alpha': 100.0,
 'class_weight': 'balanced',
 'loss': 'hinge',
 'max_iter': 1000,
 'tol': 0.001,
 'warm_start': True}

Looks like a linear SVC (`loss = 'hinge'`) with high regularization (`alpha = 100.0`) works best.

### Test set evaluation

Make test set predictions.

In [28]:
y_pred = grid.predict(X_test)

**Evaluation metrics:**

In [29]:
print(confusion_matrix(y_test, y_pred))

[[56841    19]
 [   22    80]]


Not too shabby?

In [30]:
print(classification_report(y_test, y_pred, digits=3))

             precision    recall  f1-score   support

          0      1.000     1.000     1.000     56860
          1      0.808     0.784     0.796       102

avg / total      0.999     0.999     0.999     56962



F1-score of 0.796.

In [31]:
matthews_corrcoef(y_test, y_pred)

0.7957484074536006

Informally, the model is 0.79% of the way to a perfect predictor.

The `average_precision_score` ([documentation](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html)) is an AUPRC-like metric (Area Under Precision-Recall Curve).

In [32]:
average_precision_score(y_test, y_pred)

0.6341750914768634

Not great, but we did not optimize the hyperparameters with respect to this metric.

## Random Forest Classifier
<a id='3.2'></a>

In [33]:
from sklearn.ensemble import RandomForestClassifier

### Train

The random forest takes much longer to train on this fairly large dataset, so we won't tune the hyperparameters with a grid search. We'll just use default values, except for using 100 estimators (the default is 10) and building trees in parallel (`n_jobs = -1`).

In [48]:
rf = RandomForestClassifier(n_estimators=100, n_jobs=-1)

In [49]:
rf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=-1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

### Test set evaluation

Make test set predictions.

In [50]:
y_pred = rf.predict(X_test)

**Evaluation metrics:**

In [51]:
print(confusion_matrix(y_test, y_pred))

[[56854     6]
 [   25    77]]


Looks like an improvement over the linear SVC!

In [52]:
print(classification_report(y_test, y_pred, digits=3))

             precision    recall  f1-score   support

          0      1.000     1.000     1.000     56860
          1      0.928     0.755     0.832       102

avg / total      0.999     0.999     0.999     56962



F1-score of 0.832.

In [53]:
matthews_corrcoef(y_test, y_pred)

0.8366009115262671

In [54]:
average_precision_score(y_test, y_pred, average='weighted')

0.7007696237874707

The random forest did a little better than the linear SVC.